### PR TITLE
Preprocess policy on new revision to pre-calculate hashes and minimize memory allocations.

### PR DESCRIPTION
**Pull request by scunningham**
## What does this PR do?

Rolling out a new policy caused a lot of memory and cpu thrashing. This PR moves much of the expensive calculations to happen only once on a policy change.

## Why is it important?

Fleet Server is memory bursty on policy change.  This is an issue in low memory environments.

## Checklist

- [ x] My code follows the style guidelines of this project
- [ x] I have commented my code, particularly in hard-to-understand areas
- [ x] I have made corresponding changes to the documentation
- [ x] I have made corresponding change to the default configuration files
- [ x] I have added tests that prove my fix is effective or that my feature works
- [ x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

### CODE REVIEW

File: cmd/fleet/handleCheckin.go [hunk 1]
```diff
@@ -429,7 +566,7 @@ func parsePolicy(ctx context.Context, bulker bulk.Bulk, agentId string, p model.
 
 	return &resp, nil
 }
-
+*/
```
Comment (blakerouse) [inline cmd/fleet/handleCheckin.go:569]: Should this just be removed? Why leave it in and commented out?

File: internal/pkg/policy/parsed_policy.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+type RoleMapT map[string]RoleT
+
+type ParsedPolicy struct {
+	Policy model.Policy
+	Fields map[string]json.RawMessage
+	Roles  RoleMapT
+}
+
+func NewParsedPolicy(p model.Policy) (*ParsedPolicy, error) {
```
Comment (blakerouse) [inline internal/pkg/policy/parsed_policy.go:27]: Should add unit tests for this function so we can ensure that a `model.Policy` converts to a `ParsedPolicy` correctly.

**Review by blakerouse — COMMENTED** [review]
Overall looks good, like the idea of pre-parsing the policy so its not done per Agent.

Should add some unit tests around the parsed policy.

**Review by aleksmaus — APPROVED** [review]
👍

File: cmd/fleet/handleCheckin.go (hunk 1 above)
Comment (scunningham) [reply cmd/fleet/handleCheckin.go:569]: Oversight.  Removed it; thanks.

File: internal/pkg/policy/parsed_policy.go (hunk 2 above)
Comment (scunningham) [reply internal/pkg/policy/parsed_policy.go:27]: Will do.

**Pull request merged**